internal/host/host.go

```diff
@@ -1412,3 +1418,82 @@ func (m *Manager) HandleReclaimBootArtifactDownload(ctx context.Context, h *mode
 func (m *Manager) HandleReclaimFailure(ctx context.Context, h *models.Host) error {
 	return m.sm.Run(TransitionTypeReclaimFailed, newStateHost(h), &TransitionArgsUnbindHost{ctx: ctx, db: m.db})
 }
+
+// autorenameHost renames a host if it is necessary and possible. For example, if the name is
+// localhost.localdomain and the MAC address of the main network interface is A8:CD:16:AE:79:01 it
+// will be renamed to a8-cd-16-ae-79-01. Returns a boolean indicating if the host was actually
+// renamed.
+func (m *Manager) autorenameHost(host *models.Host, inventory *models.Inventory) bool {
+	if host.RequestedHostname != "" {
+		return false
+	}
+	if inventory == nil || inventory.Hostname == "" {
+		return false
+	}
+	if !funk.Contains(hostutil.ForbiddenHostnames, inventory.Hostname) {
+		return false
+	}
+	if len(inventory.Interfaces) < 1 {
+		return false
+	}
+	mainNIC := m.mainNIC(inventory)
+	if mainNIC == nil {
+		m.log.WithFields(logrus.Fields{
+			"host": host.ID,
+			"name": inventory.Hostname,
+		}).Info(
+			"Host hasn't been automatically renamed because there is no network " +
+				"interface card with a MAC address that can be used to calculate " +
+				"the new name",
+		)
+		return false
+	}
+	host.RequestedHostname = strings.ToLower(strings.ReplaceAll(mainNIC.MacAddress, ":", "-"))
+	m.log.WithFields(logrus.Fields{
+		"host": host.ID,
+		"old":  inventory.Hostname,
+		"new":  host.RequestedHostname,
+		"nic":  mainNIC.Name,
+		"mac":  mainNIC.MacAddress,
+	}).Info("Host has been automatically renamed")
+	return true
+}
+
+// mainNIC returns the main network interface of the given host inventory, or nil if there is no
+// such network interface card. The main network interface card is the first that has assigned both
+// a MAC address and an IP address that is not local.
+func (m *Manager) mainNIC(inventory *models.Inventory) *models.Interface {
+	for _, candidate := range inventory.Interfaces {
+		hasMAC := candidate.MacAddress != ""
+		hasV4 := false
+		for _, ip := range candidate.IPV4Addresses {
+			if !m.isLocalIP(ip) {
+				hasV4 = true
+				break
+			}
+		}
+		hasV6 := false
+		for _, ip := range candidate.IPV6Addresses {
+			if !m.isLocalIP(ip) {
+				hasV6 = true
+				break
+			}
+		}
+		if hasMAC && (hasV4 || hasV6) {
+			return candidate
+		}
+	}
+	return nil
+}
+
+// isLocalIP determines if the given IP address is a local one, like 127.0.0.1 or ::1.
+func (m *Manager) isLocalIP(ip string) bool {
+	parsed := net.ParseIP(ip)
+	if parsed == nil {
+		m.log.WithFields(logrus.Fields{
+			"ip": ip,
+		}).Info("Failed to parse IP address, will assume that isn't local")
+		return false
```

Comment:
Sorry this logic is a little confusing to me, if it's an IP that can't be parsed, shouldn't it fail the condition? Since the condition used looks for a `false` in order to set the IP https://github.com/openshift/assisted-service/pull/4451/files#diff-549fd6f951612ce99043912371fffacd86d8249a40ef3905825be3d45f81dba4R1477

Reply:
This methods returns `true` if the IP is local (like `127.0.0.1` or `::1`) and `false` otherwise. For unparseable IP addresses, for example `junk`, it returns `false`. The logic being that `junk` is not `127.0.0.1` or `::1`. It is indeed weak logic because `junk` it isn't a valid IP address at all. But it doesn't affect much the rest of the system because I assume that we will never have an unparseable IP address in the inventory. I will rework this to return an error in that case.

Reply:
@CrystalChun the new version of the patch handles the non parseable addresses returning an error. Hope that makes it clearer.

Reply:
Looks good! Thanks Juan!